Genie Changes for sbndcode

nathanielerowe opened the pull request:
Hi fnal_art team,

This PR is the first in a series consisting of changes I had made to build sbndcode bare-metal on ALCF's polaris. These changes focus on changes to genie and genie-adjacent packages. The most significant changes included here are from updating genie-xsec to include tunes.

Additionally, I wanted to point out a peculiar issue I ran into on both polaris and the sbnd gpvm's (AL9) at Fermilab, which is an issue with some undefined references.

```
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: cxpm.o: in function `ErrorMessage':
/tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:116:(.text+0x25): undefined reference to `libintl_gettext'
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: /tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:105:(.text+0x48): undefined reference to `libintl_gettext'
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: /tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:110:(.text+0x5a): undefined reference to `libintl_gettext'
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: /tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:128:(.text+0x6c): undefined reference to `libintl_gettext'
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: /tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:132:(.text+0xa2): undefined reference to `libintl_gettext'
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: cxpm.o:/tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:121: more undefined references to `libintl_gettext' follow
/exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/bin/ld: cxpm.o: in function `main':
/tmp/nrowe/spack-stage/spack-stage-libxpm-3.5.17-hhytvw4dlhzfxwgfuplhc5ed5kow2dne/spack-src/cxpm/cxpm.c:149:(.text.startup+0x2c): undefined reference to `libintl_bindtextdomain'

```

The full spack-build-out and spec associated with this build (before adding the ~x) are attached. This seems very similar to this upstream issue https://github.com/spack/spack/issues/38462 . The ~x added to root in the genie package.py is meant to be a temporary fix until this issue is fixed.

I have not submitted a PR to a spack recipe repo before, so if there is anything I should provide that I did not, please let me know. I have created this PR as a draft with the intent of getting feedback on some of the things implemented here before merging.

[spack-build-out.txt](https://github.com/user-attachments/files/16567067/spack-build-out.txt)
[spack-spec.txt](https://github.com/user-attachments/files/16567068/spack-spec.txt)

Thanks for your time,
Nathaniel

Commit:
Add new genie version

Commit:
Add genie-xsec, fix genie install, sbndcode paths

nathanielerowe committed:
Create README.md

Commit:
cxxstd fix

Commit:
libxpm temp fix

nathanielerowe committed:
Update package.py

Typo in urlbase

gartung commented on packages/genie/package.py:
```diff
@@ -69,7 +70,13 @@ def url_for_version(self, version):
-    depends_on("root+pythia6")
+    # Issues caused by default root cxxstd being 11
+    # Separate issue caused by libxmp from +x requirement
+    depends_on("root @6.24.08: +pythia6 ~x cxxstd=11", when="cxxstd=11")
```
The pythia6 variant is removed in root 6.30.

nathanielerowe replied:
Ok, I can add a restriction on the versioning such that it does not apply that variant for 6.30 and beyond. Thanks!

nathanielerowe replied:
Using newer versions of root seems to not work with genie due to pythia6 issues. I am going to just restrict the version of root (for now) in the genie package.py. Looks like header files were removed from root at some point, similar to a larsim issue I submitted last week on redmine. Is this an issue that should be handed off to genie people? Does this fix seem reasonable?

[spack-build-out.txt](https://github.com/user-attachments/files/16714553/spack-build-out.txt)

gartung replied:
Seems OK to me.

gartung commented on packages/lhapdf/package.py:
```diff
@@ -51,6 +51,8 @@ def patch(self):
 
     depends_on("pdfsets")
+    depends_on("python")
+    depends_on("gettext")
```
What are python and gettext needed for?

nathanielerowe replied:
Adding python fixed the following error for me:

```
==> Error: ProcessError: Command exited with status 1:
    '/tmp/nrowe/spack-stage/spack-stage-lhapdf-6.5.3-xq4yzkqahcszya6isvzlf23mzumbjaw2/spack-src/configure' '--prefix=/exp/sbnd/app/users/nrowe/spack_dev/spack/opt/spack/linux-almalinux9-zen3/gcc-11.4.1/lhapdf-6.5.3-xq4yzkqahcszya6isvzlf23mzumbjaw2' '--enable-low-memory' '--disable-pyext' '--disable-octave'

2 errors found in build log:
     112    checking for Python library path... -L/exp/sbnd/app/users/nrowe/spack_dev/spack/opt/spack/linux-almalinux9-x86_64_v3/gcc-11.4.1/python-3.11.9-ptsa74fzbyct53xevm6ffevlyqtyagsd/lib -lpython3.11
     113    checking for Python site-packages path... /exp/sbnd/app/users/nrowe/spack_dev/spack/var/spack/environments/testing/.spack-env/view/lib/python3.11/site-packages
     114    checking for Python platform specific site-packages path...
     115    checking python extra libraries... -lintl -ldl -lm
     116    checking python extra linking flags... -Xlinker -export-dynamic
     117    checking consistency of all components of python development environment... no
  >> 118    configure: error: in `/tmp/nrowe/spack-stage/spack-stage-lhapdf-6.5.3-xq4yzkqahcszya6isvzlf23mzumbjaw2/spack-src':
  >> 119    configure: error:
     120      Could not link test program to Python. Maybe the main Python library has been
     121      installed in some non-standard library path. If so, pass it to configure,
     122      via the LIBS environment variable.
     123      Example: ./configure LIBS="-L/usr/non-standard-path/python/lib"
     124      ============================================================================
     125       ERROR!
```
[python-spack-build-out.txt](https://github.com/user-attachments/files/16604660/python-spack-build-out.txt)

This issues seem somewhat related to [https://github.com/spack/spack/pull/37384](url). Since there isn't a python variant in this version of the package.py, it may just be assuming that variant by default since it is detecting a python already in the genie build. 

It may be easier to just update the whole lhapdf package.py to match the nominal spack version. I also noticed that lhapdf is already included in both the nominal spack and the fermilab fork [https://github.com/FNALssi/spack](url). This also seems to be true for genie itself. Am I using the preferred repos for these package.py files? Do these different versions serve different purposes?

Previously I was getting an additional error associated with not including gettext only on polaris. I am no longer getting that error if I include only the python, so I will remove that line and let you know if that becomes an issue again.

gartung replied:
This version of lhapdf was created before there was a version in mainline spack. If they are similar enough we could remove this one from fnal_art.

nathanielerowe replied:
I will try installing with the mainline spack version and make sure it works before removing the file from the PR. I will follow up on this.

nathanielerowe replied:
Using the lhapdf in the fermi spack fork works for lhapdf@6.5.4~python  lhapdf@6.5.3 but not lhapdf@6.5.4+python. It returns the following errors

```
make[1]: Entering directory '/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/wrappers'
Making all in python
make[2]: Entering directory '/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/wrappers/python'
cython ./lhapdf.pyx --cplus -2 -o lhapdf.cpp
for i in lhapdf.cpp lhapdf.pyx clhapdf.pxd __init__.py; do \
  orig="$i"; \
  build="$(basename $i)"; \
  if [ ! -e ${build} ]; then cp ${orig} ${build}; \
  echo ${build} >> fix-out-of-source; fi; \
done
touch fix-out-of-source
CC="/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/lib/spack/env/gcc/g++" CXX="/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/lib/spack/env/gcc/g++" CFLAGS="-O3" CXXFLAGS="-O3" CPPFLAGS="-I/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/include -I." LHAPDF_LOCAL=1 /lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/python-venv-1.0-d6kjr2thd5iybzcu3jjk6nfvrip5yrrs/bin/python build.py
/usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: cannot find -lintl: No such file or directory
collect2: error: ld returned 1 exit status
Build command = /lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/lib/spack/env/gcc/g++  -shared -fPIC  -o  lhapdf.so  /var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/wrappers/python/lhapdf.cpp  -I/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/include -I/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/include -I/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/include  -I/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/include   -O3   -L/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/src/.libs  -L/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/lib  -lstdc++ -lLHAPDF  -I/lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/python-3.11.9-yajs77zqrlbvpsyqicvhgxme2ymdz6vj/include/python3.11 /lus/grand/projects/neutrinoGPU/software/NateDev/fermi-spack-dev/spack/opt/spack/linux-sles15-zen3/gcc-12.3.0/python-3.11.9-yajs77zqrlbvpsyqicvhgxme2ymdz6vj/lib/libpython3.11.so -lintl -lpthread -ldl  -lutil -lm
make[2]: *** [Makefile:499: all-local] Error 1
make[2]: Leaving directory '/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/wrappers/python'
make[1]: *** [Makefile:383: all-recursive] Error 1
make[1]: Leaving directory '/var/tmp/pbs.2064452.polaris-pbs-01.hsn.cm.polaris.alcf.anl.gov/nathanielerowe/spack-stage/spack-stage-lhapdf-6.5.4-w4was2bjecpv3kld7sd2gfgub3b5dcjf/spack-src/wrappers'
make: *** [Makefile:520: all-recursive] Error 1
```

[spack-build-out.txt](https://github.com/user-attachments/files/16619725/spack-build-out.txt)

I was able to get rid of these errors by updating the gettext package.py in the fermi spack fork to match the main line spack. After this, genie is able to build without the lhapdf in fnal_art but with the one in the fermi spack fork. Should I just raise an issue on the fermi spack fork regarding this?

gartung replied:
Yes, please make an issue. We should switch to using the upstream lhapdf and gettext.

nathanielerowe replied:
Made a discussion [here](https://github.com/FNALssi/spack/discussions/423) as I don't see an issues tab.

Commit:
remove lhapdf, root vers, misc fixes

Commit:
Undo some changes (for now)

nathanielerowe committed:
Delete README.md

nathanielerowe committed:
Update package.py

Adding back in text block

nathanielerowe committed:
restrict root version

gartung commented on packages/genie/package.py:
```diff
@@ -67,9 +68,16 @@ def url_for_version(self, version):
 
-    depends_on("root+pythia6")
+    # Issues caused by default root cxxstd being 11
+    # Separate issue caused by libxmp from +x requirement
```
The `~x` variant on `root` should be applied externally via command line or environment requirements.

nathanielerowe committed:
Update package.py

Remove ~x variant. libxpm build issue resolved by updating gettext to match mainline spack.
